File: Year2023/CSharp/Day07.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace AdventOfCode.Year2023
{
    // https://www.reddit.com/r/adventofcode/comments/18cq5j3/2023_day_7_part_1_two_hints_for_anyone_stuck/
    // https://www.reddit.com/r/adventofcode/comments/18cr4xr/2023_day_7_better_example_input_not_a_spoiler/

    [TestFixture]
    [NonParallelizable]
    public sealed class Day07
    {
        readonly string[] Input;

        struct Hand
        {
            public HandType Type { get; set; }
            public byte[] CardValues { get; set; }
            public short Bid { get; set; }
            public string CardString { get; set; }
        }

        enum HandType : byte
        {
            NONE = 0,
            FiveOfaKind = 7,
            FourOfaKind = 6,
            FullHouse = 5,
            ThreeOfaKind = 4,
            TwoPair = 3,
            OnePair = 2,
            HighCard = 1
        }

        Dictionary<char, byte> CardValues;

        readonly Hand[] Hands;

        public Day07()
        {
            Input = File.ReadAllLines("Input/07.txt");

            // Part 1 test data
            /*Input = new string[]
            {
                "2345A 1",      // 00 High card
                "Q2KJJ 13",     // 01 One pair
                "Q2Q2Q 19",     // 02 Full house
                "T3T3J 17",     // 03 Two pairs
                "T3Q33 11",     // 04 Three of a kind
                "2345J 3",      // 05 High card
                "J345A 2",      // 06 High card
                "32T3K 5",      // 07 One pair
                "T55J5 29",     // 08 Three of a kind
                "KK677 7",      // 09 Two pairs
                "KTJJT 34",     // 10 Two pairs
                "QQQJA 31",     // 11 Three of a kind
                "JJJJJ 37",     // 12 Five of a kind
                "JAAAA 43",     // 13 Four of a kind
                "AAAAJ 59",     // 14 Four of a kind
                "AAAAA 61",     // 15 Five of a kind
                "2AAAA 23",     // 16 Four of a kind
                "2JJJJ 53",     // 17 Four of a kind
                "JJJJ2 41"      // 18 Four of a kind
            };*/

            // Part 1 = 6592 for test input

            Hands = new Hand[Input.Length];
        }

        [TestCase]
        public void Part01()
        {
            CardValues = new()
            {
                { 'A', 13 },
                { 'K', 12 },
                { 'Q', 11 },
                { 'J', 10 },
                { 'T', 9 },
                { '9', 8 },
                { '8', 7 },
                { '7', 6 },
                { '6', 5 },
                { '5', 4 },
                { '4', 3 },
                { '3', 2 },
                { '2', 1 }
            };

            int HandInsertIndex = 0;

            Span<byte> CardBuffer = stackalloc byte[13];
            Span<byte> Cards = stackalloc byte[5];

            foreach (string CurrentInput in Input)
            {
                ReadOnlySpan<char> CardLettersSlice = CurrentInput.AsSpan(0, 5);

                for(int Index = 0; Index < CardLettersSlice.Length; Index++)
                {
                    if (CardLettersSlice[Index] == 'A') CardBuffer[12]++;
                    else if (CardLettersSlice[Index] == 'K') CardBuffer[11]++;
                    else if (CardLettersSlice[Index] == 'Q') CardBuffer[10]++;
                    else if (CardLettersSlice[Index] == 'J') CardBuffer[9]++;
                    else if (CardLettersSlice[Index] == 'T') CardBuffer[8]++;
                    else if (CardLettersSlice[Index] == '9') CardBuffer[7]++;
                    else if (CardLettersSlice[Index] == '8') CardBuffer[6]++;
                    else if (CardLettersSlice[Index] == '7') CardBuffer[5]++;
                    else if (CardLettersSlice[Index] == '6') CardBuffer[4]++;
                    else if (CardLettersSlice[Index] == '5') CardBuffer[3]++;
                    else if (CardLettersSlice[Index] == '4') CardBuffer[2]++;
                    else if (CardLettersSlice[Index] == '3') CardBuffer[1]++;
                    else if (CardLettersSlice[Index] == '2') CardBuffer[0]++;

                    Cards[Index] = CardValues[CardLettersSlice[Index]];
                }

                var NewHand = new Hand()
                {
                    CardValues = Cards.ToArray(),
                    Bid = Convert.ToInt16(CurrentInput[6..])
                };

                HandType HandType = HandType.NONE;
                bool AnyCountOf3 = false;
                bool AnyCountOf2 = false;
                int CountsOfOne = 0;

                for (int Index = 0; Index < CardBuffer.Length; Index++)
                {
                    byte CurrentCardCount = CardBuffer[Index];
                    if (CurrentCardCount == 5)
                    {
                        HandType = HandType.FiveOfaKind;
                        break;
                    }

                    if (CurrentCardCount == 4)
                    {
                        HandType = HandType.FourOfaKind;
                        break;
                    }

                    if (CurrentCardCount == 3) AnyCountOf3 = true;
                    else if (CurrentCardCount == 2) AnyCountOf2 = true;
                    else if (CurrentCardCount == 1) CountsOfOne++;
                }

                if (HandType == HandType.NONE)
                {
                    if (AnyCountOf3 && AnyCountOf2) HandType = HandType.FullHouse;
                    else if (AnyCountOf3 && CountsOfOne > 0) HandType = HandType.ThreeOfaKind;
                    else if (AnyCountOf2 && CountsOfOne == 1) HandType = HandType.TwoPair;
                    else if (AnyCountOf2 && CountsOfOne == 3) HandType = HandType.OnePair;
                    else HandType = HandType.HighCard;
                }

                NewHand.Type = HandType;
                NewHand.CardString = new string(CardLettersSlice);

                CardBuffer.Clear();
                Cards.Clear();

                Hands[HandInsertIndex] = NewHand;
                HandInsertIndex++;
            }

            long FinalValue = 0;

            List<Hand> HandsOrdered = Hands.ToList();
            HandsOrdered.Sort((x, y) =>
            {
                byte xTypeValue = (byte)x.Type;
                byte yTypeValue = (byte)y.Type;
                if (xTypeValue < yTypeValue) return -1;
                if (xTypeValue > yTypeValue) return 1;

                for (int Index = 0; Index < x.CardValues.Length; Index++)
                {
                    if (x.CardValues[Index] < y.CardValues[Index]) return -1;
                    if (x.CardValues[Index] > y.CardValues[Index]) return 1;
                }

                return 0;
            });

            for(int Index = Hands.Length - 1; Index != -1; Index--)
            {
                var CurrentHand = HandsOrdered[Index];
                FinalValue += CurrentHand.Bid * (Index + 1);
            }

            Console.WriteLine($"Part one answer: " + FinalValue);
            Trace.Assert(FinalValue == 246424613, $"Expected answer to be 246424613 but it was {FinalValue}");
        }

        //[TestCase]
        public void Part02()
        {
            CardValues = new()
            {
                { 'A', 12 },
                { 'K', 11 },
                { 'Q', 10 },
                { 'T', 9 },
                { '9', 8 },
                { '8', 7 },
                { '7', 6 },
                { '6', 5 },
                { '5', 4 },
                { '4', 3 },
                { '3', 2 },
                { '2', 1 },
                { 'J', 0 }
            };

            throw new NotImplementedException("Not finished");
        }
    }
}

File: Year2023/C/src/Day07.c
// #include "LibNumeric.h"
#include "LibString.h"
#include "LibThomas.h"
#include "LibString.h"
#include "stdio.h"
#include <assert.h>
#include <limits.h>
#include <pthread.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

typedef enum {
    HandType_NONE = 0,
    HandType_FiveOfaKind = 7,
    HandType_FourOfaKind = 6,
    HandType_FullHouse = 5,
    HandType_ThreeOfaKind = 4,
    HandType_TwoPair = 3,
    HandType_OnePair = 2,
    HandType_HighCard = 1
} HandType;

typedef struct _Hand {
    HandType Type;
    int32_t Bid;
    char Cards[5];
} Hand;

int32_t DetermineCardValue(const char input)
{
    switch (input)
    {
        case 'A': return 13;
        case 'K': return 12;
        case 'Q': return 11;
        case 'J': return 10;
        case 'T': return 9;
        case '9': return 8;
        case '8': return 7;
        case '7': return 6;
        case '6': return 5;
        case '5': return 4;
        case '4': return 3;
        case '3': return 2;
        case '2': return 1;
        default: return 0;
    }
}

Hand* ParseCardIntoHand(const String *input)
{
    int32_t CardCount[13] = { 0 }; 
    
    for (int32_t index = 0; index < 5; index++) {

        char NextChar = input->Content[index];

        switch (NextChar) {
            case 'A': CardCount[0]++; break;
            case 'K': CardCount[1]++; break;
            case 'Q': CardCount[2]++; break;
            case 'J': CardCount[3]++; break;
            case 'T': CardCount[4]++; break;
            case '9': CardCount[5]++; break;
            case '8': CardCount[6]++; break;
            case '7': CardCount[7]++; break;
            case '6': CardCount[8]++; break;
            case '5': CardCount[9]++; break;
            case '4': CardCount[10]++; break;
            case '3': CardCount[11]++; break;
            case '2': CardCount[12]++; break;
        }
    }

    Hand *ReturnData = malloc(sizeof(Hand) + sizeof(char[5]));
    ReturnData->Bid = 0;
    ReturnData->Type = HandType_NONE;

    bool TwoOfAKind = false;
    bool ThreeOfAKind = false;
    int32_t Singles = 0;

    for (int32_t index = 0; index < 13; index++) {
        int32_t Current = CardCount[index];

        if (Current == 5) {
            ReturnData->Type = HandType_FiveOfaKind;
            break;
        }

        if (Current == 4) {
            ReturnData->Type = HandType_FourOfaKind;
            break;
        }

        if (Current == 1) {
            Singles++;
            if (Singles == 5) {
                ReturnData->Type = HandType_HighCard;
                break;
            }
        }

        if (Current == 3) {

            if (TwoOfAKind) {
                ReturnData->Type = HandType_FullHouse;
                break;
            }

            ThreeOfAKind = true;
        }

        if (Current == 2) {

            if (TwoOfAKind) {
                ReturnData->Type = HandType_TwoPair;
                break;
            }

            if (ThreeOfAKind) {
                ReturnData->Type = HandType_FullHouse;
                break;
            }

            TwoOfAKind = true;
        }

        if (ThreeOfAKind && Singles == 2) {
            ReturnData->Type = HandType_ThreeOfaKind;
            break;
        }

        if (TwoOfAKind && Singles == 3) {
            ReturnData->Type = HandType_OnePair;
            break;
        }
    }

#if DEBUG()
    assert(ReturnData->Type != HandType_NONE);
#endif

    int32_t Bid = 0;
    for(int32_t index = 6; index < input->Size; index++) {
        Bid = Bid * 10;
        Bid = Bid + (input->Content[index] - '0');
    }

    ReturnData->Bid = Bid;
    memcpy(ReturnData->Cards, input->Content, 5);

    return ReturnData;
}

int SortHands(const void *elem1, const void *elem2)
{
    Hand *Value1 = (Hand*)elem1;
    Hand *Value2 = (Hand*)elem2;

    if (Value1->Type < Value2->Type) return -1;
    if (Value1->Type > Value2->Type) return 1;

    for (int Index = 0; Index < 5; Index++)
    {
        if (DetermineCardValue(Value1->Cards[Index]) < DetermineCardValue(Value2->Cards[Index])) return -1;
        if (DetermineCardValue(Value1->Cards[Index]) > DetermineCardValue(Value2->Cards[Index])) return 1;
    }

    return 0;
}

int main(void)
{
    int32_t Part1Answer = 0;
    
    /*String *TestString[19] = {0};
    TestString[0] = &(String){ .Size = 8, .Content = "JJJJJ 37" }; // 50: Five of a kind
    TestString[1] = &(String){ .Size = 8, .Content = "AAAAA 61" }; // 65: Five of a kind
    TestString[2] = &(String){ .Size = 8, .Content = "JAAAA 43" }; // 62: Four of a kind
    TestString[3] = &(String){ .Size = 8, .Content = "AAAAJ 59" }; // 62: Four of a kind
    TestString[4] = &(String){ .Size = 8, .Content = "2AAAA 23" }; // 53: Four of a kind
    TestString[5] = &(String){ .Size = 8, .Content = "2JJJJ 53" }; // 41: Four of a kind
    TestString[6] = &(String){ .Size = 8, .Content = "JJJJ2 41" }; // 41: Four of a kind
    TestString[7] = &(String){ .Size = 8, .Content = "Q2Q2Q 19" }; // 35: Full house
    TestString[8] = &(String){ .Size = 8, .Content = "T3Q33 11" }; // 26: Three of a kind
    TestString[9] = &(String){ .Size = 8, .Content = "T55J5 29" }; // 31: Three of a kind
    TestString[10] = &(String){ .Size = 8, .Content = "QQQJA 31" };// 56: Three of a kind
    TestString[11] = &(String){ .Size = 8, .Content = "T3T3J 17" };// 32: Two pairs
    TestString[12] = &(String){ .Size = 7, .Content = "KK677 7" }; // 41: Two pairs
    TestString[13] = &(String){ .Size = 8, .Content = "KTJJT 34" };// 50: Two pairs
    TestString[14] = &(String){ .Size = 8, .Content = "Q2KJJ 13" };// 44: One pair
    TestString[15] = &(String){ .Size = 7, .Content = "32T3K 5" }; // 26: One pair
    TestString[16] = &(String){ .Size = 7, .Content = "2345A 1" }; // 23: High card
    TestString[17] = &(String){ .Size = 7, .Content = "2345J 3" }; // 20: High card
    TestString[18] = &(String){ .Size = 7, .Content = "J345A 2" }; // 32: High card
    StringArray *TestInput = StringArray_Make(TestString, sizeof(TestString) / sizeof(TestString[0]));*/

    StringArray *TestInput = File_ReadAllLines("Input/07.txt");

    Hand Hands[TestInput->Count];
    memset(Hands, 0, sizeof(Hand));

#if DEBUG()
    assert(sizeof(Hands) / sizeof(Hands[0]) == TestInput->Count);
#endif

    // PART 01
    for (int32_t index = 0; index < TestInput->Count; index++) {
        
        String *CurrentInput = TestInput->Contents[index];
        Hand *TheHand = ParseCardIntoHand(CurrentInput);
        Hands[index] = *TheHand;
    }

    qsort(Hands, TestInput->Count, sizeof(Hands[0]), SortHands);

    for(int Index = TestInput->Count - 1; Index != -1; Index--)
    {
        Hand CurrentHand = Hands[Index];
        Part1Answer += CurrentHand.Bid * (Index + 1);
    }
    // END PART 01

    StringArray_Free(TestInput, true);

    printf("Part 01 answer: %i\n", Part1Answer);
    assert(Part1Answer == 246424613);
    // assert(Part1Answer == 6592);
}
